core/src/cluster_slots.rs

```diff
@@ -0,0 +1,267 @@
+use crate::{cluster_info::ClusterInfo, epoch_slots::EpochSlots, serve_repair::RepairType};
+
+use solana_ledger::{bank_forks::BankForks, staking_utils};
+use solana_sdk::{clock::Slot, pubkey::Pubkey};
+
+use std::{
+    collections::{HashMap, HashSet},
+    rc::Rc,
+    sync::RwLock,
+};
+
+#[derive(Default)]
+pub struct ClusterSlots {
+    cluster_slots: HashMap<Slot, HashMap<Rc<Pubkey>, u64>>,
+    keys: HashSet<Rc<Pubkey>>,
+    since: Option<u64>,
+    default_peers: HashMap<Rc<Pubkey>, u64>,
+    epoch: Option<u64>,
+    self_id: Pubkey,
+}
+
+impl ClusterSlots {
+    pub fn lookup(&self, slot: Slot) -> Option<&HashMap<Rc<Pubkey>, u64>> {
+        self.cluster_slots.get(&slot)
+    }
+    pub fn update(
+        &mut self,
+        root: Slot,
+        cluster_info: &RwLock<ClusterInfo>,
+        bank_forks: &RwLock<BankForks>,
+    ) {
+        self.update_peers(cluster_info, bank_forks);
+        let epoch_slots = cluster_info
+            .read()
+            .unwrap()
+            .get_epoch_slots_since(self.since);
+        self.update_internal(root, epoch_slots);
+    }
+    fn update_internal(&mut self, root: Slot, epoch_slots: (Vec<EpochSlots>, Option<u64>)) {
+        let (epoch_slots_list, since) = epoch_slots;
+        for epoch_slots in epoch_slots_list {
+            let slots = epoch_slots.to_slots(root);
+            for slot in &slots {
+                if *slot <= root {
+                    continue;
+                }
+                let pubkey = Rc::new(epoch_slots.from);
+                if self.keys.get(&pubkey).is_none() {
+                    self.keys.insert(pubkey.clone());
+                }
+                let from = self.keys.get(&pubkey).unwrap();
+                let balance = self.default_peers.get(from).cloned().unwrap_or(0);
+                if self.self_id != **from {
+                    debug!(
+                        "CLUSTER_SLLOTS: {}: insert {} {} {}",
+                        self.self_id, from, *slot, balance
+                    );
+                }
+                self.cluster_slots
+                    .entry(*slot)
+                    .or_insert_with(HashMap::default)
+                    .insert(from.clone(), balance + 1);
```

Comment:
hmmm why `balance + 1`?

Reply:
Weights fail over with 0 weight.  I think I don’t need to do this here.